File: tools/mlflow_fun/tools/dump_run.py
"""
Run dump utilities.
"""

import time
import mlflow

INDENT = "  "
MAX_LEVEL = 1
TS_FORMAT = "%Y-%m-%d_%H:%M:%S"
client = mlflow.tracking.MlflowClient()

def dump_run(run, max_level=1, indent=""):
    dump_run_info(run.info,indent)
    print(indent+"Params:")
    for k,v in sorted(run.data.params.items()):
        print(indent+"  {}: {}".format(k,v))
    print(indent+"Metrics:")
    for k,v in sorted(run.data.metrics.items()):
        print(indent+"  {}: {}".format(k,v))
    print(indent+"Tags:")
    for k,v in sorted(run.data.tags.items()):
        print(indent+"  {}: {}".format(k,v))
    print("{}Artifacts:".format(indent))
    dump_artifacts(run.info.run_id, "", 0, max_level, indent+INDENT)
    return run
        
def dump_run_id(run_id, max_level=1, indent=""):
    run = client.get_run(run_id)
    return dump_run(run,max_level,indent)

def dump_run_info(info, indent=""):
    print("{}RunInfo:".format(indent))
    for k,v in sorted(info.__dict__.items()):
        if not k.endswith("_time"):
            print("{}  {}: {}".format(indent,k[1:],v))
    start = _dump_time(info,'_start_time',indent)
    end = _dump_time(info,'_end_time',indent)
    if start is not None and end is not None:
        dur = float(end - start)/1000
        print("{}  _duration:  {} seconds".format(indent,dur))

def _dump_time(info, k, indent=""):
    v = info.__dict__.get(k,None)
    if v is None:
        print("{}  {:<11} {}".format(indent,k[1:]+":",v))
    else:
        stime = time.strftime(TS_FORMAT,time.gmtime(v/1000))
        print("{}  {:<11} {}   {}".format(indent,k[1:]+":",stime,v))
    return v

def dump_artifacts(run_id, path, level, max_level, indent):
    if level+1 > max_level: return
    artifacts = client.list_artifacts(run_id,path)
    for art in artifacts:
        print("{}Artifact - level {}:".format(indent,level))
        for k,v in art.__dict__.items(): print("  {}{}: {}".format(indent,k[1:],v))
        if art.is_dir:
            dump_artifacts(run_id, art.path, level+1, max_level, indent+INDENT)

if __name__ == "__main__":
    from argparse import ArgumentParser
    parser = ArgumentParser()
    parser.add_argument("--run_id", dest="run_id", help="Run ID", required=True)
    parser.add_argument("--artifact_max_level", dest="artifact_max_level", help="Number of artifact levels to recurse", required=False, default=1, type=int)
    args = parser.parse_args()
    print("args:",args)
    dump_run_id(args.run_id, args.artifact_max_level)

File: examples/scala/src/main/scala/org/andre/mlflow/tools/DumpRun.scala
package org.andre.mlflow.tools

import java.util.Date
import scala.collection.JavaConversions._
import com.beust.jcommander.{JCommander, Parameter}
import org.mlflow.tracking.MlflowClient
import org.andre.mlflow.util.MLflowUtils
import org.mlflow.api.proto.Service.RunInfo

object DumpRun {
  val FORMAT = new java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss")

  def dumpRun(client: MlflowClient, runId: String, artifactMaxLevel: Int, indent: String = "") {
    val run = client.getRun(runId)
    dumpRunInfo(run.getInfo, indent)
    println(s"${indent}Params:")
    for (x <- run.getData.getParamsList) {
      println(s"${indent}  ${x.getKey}: ${x.getValue}")
    }
    println(s"${indent}Metrics:")
    for (x <- run.getData.getMetricsList) {
      println(s"${indent}  ${x.getKey}: ${x.getValue} - ${x.getTimestamp}")
    }
    println(s"${indent}Tags:")
    for (x <- run.getData.getTagsList) {
      println(s"${indent}  ${x.getKey}: ${x.getValue}")
    }
    println(s"${indent}Artifacts:")
    dumpArtifacts(client, runId, "", 0, artifactMaxLevel, indent+"  ")
  }

  def dumpRunInfo(info: RunInfo, indent: String = "") {
    println(s"${indent}RunInfo:")
    println(s"$indent  runId: ${info.getRunId}")
    println(s"$indent  experimentId: ${info.getExperimentId}")
    println(s"$indent  lifecycleStage: ${info.getLifecycleStage}")
    println(s"$indent  userId: ${info.getUserId}")
    println(s"$indent  status: ${info.getStatus}")
    println(s"$indent  artifactUri: ${info.getArtifactUri}")
    println(s"$indent  startTime: ${info.getStartTime}")
    println(s"$indent  endTime:   ${info.getEndTime}")
    println(s"$indent  startTime: ${FORMAT.format(new Date(info.getStartTime))}")
    println(s"$indent  endTime:   ${FORMAT.format(new Date(info.getEndTime))}")
    val duration = (info.getEndTime - info.getStartTime).toDouble / 1000
    println(s"$indent  _duration: ${duration} seconds")
  }

  def dumpArtifacts(client: MlflowClient, runId: String, path: String, level: Int, maxLevel: Int, indent: String) {
    if (level+1 > maxLevel) {
       return
    }
    val artifacts = client.listArtifacts(runId, path)
    for ((art,j) <- artifacts.zipWithIndex) {
      println(s"${indent}Artifact ${j+1}/${artifacts.size} - level $level")
      println(s"$indent  path: ${art.getPath}")
      println(s"$indent  isDir: ${art.getIsDir}")
      if (art.getIsDir) {
        dumpArtifacts(client, runId, art.getPath, level+1, maxLevel, indent+"  ")
      } else {
        println(s"$indent  fileSize: ${art.getFileSize}")
      }
    }
  }

  def main(args: Array[String]) {
    new JCommander(opts, args.toArray: _*)
    println("Options:")
    println(s"  tracking URI: ${opts.trackingUri}")
    println(s"  token: ${opts.token}")
    println(s"  runId: ${opts.runId}")
    println(s"  artifactMaxLevel: ${opts.artifactMaxLevel}")
    val client = MLflowUtils.createMlflowClient(opts.trackingUri, opts.token)
    dumpRun(client, opts.runId, opts.artifactMaxLevel)
  }

  object opts {
    @Parameter(names = Array("--trackingUri" ), description = "Tracking Server URI", required=false)
    var trackingUri: String = null

    @Parameter(names = Array("--token" ), description = "REST API token", required=false)
    var token: String = null

    @Parameter(names = Array("--runId" ), description = "runId", required=true)
    var runId: String = null

    @Parameter(names = Array("--artifactMaxLevel" ), description = "Number of artifact levels to recurse", required=false)
    var artifactMaxLevel = 1
  }
}
